driver/src/price_estimation/dexag/api.rs

```diff
@@ -0,0 +1,167 @@
+#![allow(dead_code)]
+
+use anyhow::{anyhow, Context, Result};
+use ethcontract::Address;
+use isahc::prelude::*;
+use serde::{de::Error as _, Deserialize, Deserializer};
+use serde_with::rust::display_fromstr;
+use std::time::Duration;
+use url::Url;
+
+#[cfg_attr(test, mockall::automock)]
+/// Parts of the dex.ag api.
+pub trait DexagApi {
+    /// https://docs.dex.ag/api/tokens
+    fn get_token_list(&self) -> Result<Vec<Token>>;
+    /// https://docs.dex.ag/api/price
+    fn get_price(&self, from: &Token, to: &Token, amount: EitherAmount) -> Result<Price>;
+}
+
+#[derive(Clone, Debug, Deserialize)]
+pub struct Token {
+    pub name: String,
+    pub symbol: String,
+    #[serde(default)]
+    #[serde(deserialize_with = "deserialize_address")]
+    pub address: Option<Address>,
+}
+
+#[derive(Clone, Debug, Deserialize)]
+pub struct Price {
+    #[serde(with = "display_fromstr")]
+    pub price: f64,
+}
+
+fn deserialize_address<'de, D>(deserializer: D) -> Result<Option<Address>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    fn hex_string_to_address(string: &str) -> Result<Address> {
+        let prefix = "0x";
+        if !string.starts_with(prefix) {
+            return Err(anyhow!("does not start with {}", prefix));
```

Comment:
nit: Maybe it helps printing the actual string here so it's easier to debug the problem (but I see it's actually printed in the context below).